---
title: "example"
output: html_document
date: "2023-02-15"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}

#=====================================================================
# Loading data and packages -- 
#   assumes "3D_CGR/paper code" is the current working directory
#=====================================================================

source("./data/sequences.R")
source("./R/functions.R")

require(kmer)
require(grid)
require(gtable)
library(reshape2)
library(msa) # From Bioconductor
library(ape)
library(phangorn)
library(tidyverse)
library(ggplot2)
library(ggdendro)
library(ggtree)

```

```{r}
#####################################################################
# Example DNA sequences analyzed by 3D CGR / paper reproduction
# This code does not run in parallel (yet) and the section on volume intersection
# can take a while to run.
# By: Stephanie Young <syoung2@sdsu.edu> <syoung49@its.jnj.com>
#####################################################################
```

```{r}

################################################################################
# Helper functions
################################################################################

# Heatmap functions

hm <- function(distances, na.value =  rgb(0, 0, 0, alpha = .25), ...){
  if(class(distances)[1] == "dist")
    dat <- distances |> as.matrix() |> reshape2::melt(value.name = "distance") 
  else 
    dat <- distances |> reshape2::melt(value.name = "distance") 
  dat <- dat |> mutate(orig = distance) 
  dat <- dat |> mutate(distance = case_when(Var1 == Var2 ~ NA, TRUE ~ distance))
  dat <- dat |> mutate(distance = (distance - min(distance, na.rm = T)))
  dat <- dat |> mutate(distance = distance / max(distance, na.rm = T))
  #tri <- distances |> as.matrix() |> upper.tri(diag = T) |> reshape2::melt() |> pull(value) 
  #dat <- dat[tri, , drop = F] 
  dat |> 
    ggplot() + 
    geom_tile(aes(x = Var1, y = Var2, fill = distance)) + 
    viridis::scale_fill_viridis(option = "turbo", limits = c(0, 1), na.value = na.value, ... ) +
    theme_light() + 
    theme(
      axis.text.x = element_text(angle = 90, vjust = .25, hjust = 1), 
      axis.title.x = element_blank(), 
      axis.title.y = element_blank(), 
      legend.position = "bottom"
    ) +
    coord_fixed() +
    guides(fill = guide_colourbar("Normalized\ndistance"))
}

# Function to align the sequences

clustal_omega <- function(dna_seq, outgroup = NULL){
  msa(dna_seq, type = "dna", method = "ClustalOmega", order = "input") |> 
    msaConvert(type = "ape::DNAbin") |> ape::dist.gene() |> # can also use dist.p() from the phangorn package. 
    (\(x)x/max(x))() # distances rescaled to be between 0 and 1 because bionj does not like distances > 100
}

# K-mer distance function 

kmer_dist <- function(seq_string, ...){
  kmer_seq <- kmer::kdistance(str_split(seq_string, ""), ...)
  kmer_seq <- as.matrix(kmer_seq)
  colnames(kmer_seq) <- names(seq_string)
  rownames(kmer_seq) <- names(seq_string)
  kmer_seq
}

```


```{r, dpi = 400}

#=====================================================================
# Clustal Omega + p-distance results for comparison
#=====================================================================

clustal_omega(beta_seq) |> hm(trans = scales::modulus_trans(p=2)) 
clustal_omega(beta_seq) |> bionj() |> root("gallus") |> plot()
clustal_omega(beta_seq) |> upgma() |> root("gallus") |> plot()
clustal_omega(beta_seq) |> wpgma() |> root("gallus") |> plot()

clustal_omega(nadh) |> hm(trans = scales::modulus_trans(p=2)) 
clustal_omega(nadh) |> bionj() |> root("Saimiri sciureus") |> plot() # bionj does not like distances > 100
clustal_omega(nadh) |> upgma() |> root("Saimiri sciureus") |> plot()
clustal_omega(nadh) |> wpgma() |> root("Saimiri sciureus") |> plot()

clustal_omega(syn_fas) |> hm(trans = scales::modulus_trans(p=2)) 
clustal_omega(syn_fas) |> bionj() |> plot() 
clustal_omega(syn_fas) |> upgma() |> plot() 
clustal_omega(syn_fas) |> wpgma() |> plot() 

#=====================================================================
# kmer-distance results for comparison 
#=====================================================================


kmer_dist(beta_seq, k = 6, method = "euclidean") |> hm(trans = scales::modulus_trans(p=2)) 
kmer_dist(beta_seq, k = 6, method = "euclidean") |> bionj() |> root("gallus") |> plot()
kmer_dist(beta_seq, k = 6, method = "euclidean") |> upgma() |> root("gallus") |> plot()
kmer_dist(beta_seq, k = 6, method = "euclidean") |> wpgma() |> root("gallus") |> plot()

kmer_dist(nadh, k = 6, method = "euclidean") |> hm(trans = scales::modulus_trans(p=2))
kmer_dist(nadh, k = 6, method = "euclidean") |> bionj() |> root("Saimiri sciureus") |> plot()
kmer_dist(nadh, k = 6, method = "euclidean") |> upgma() |> root("Saimiri sciureus") |> plot()
kmer_dist(nadh, k = 6, method = "euclidean") |> wpgma() |> root("Saimiri sciureus") |> plot()

kmer_dist(syn_fas, k = 6, method = "euclidean") |> hm(trans = scales::modulus_trans(p=2))
kmer_dist(syn_fas, k = 6, method = "euclidean") |> bionj() |> plot()
kmer_dist(syn_fas, k = 6, method = "euclidean") |> upgma() |> plot()
kmer_dist(syn_fas, k = 6, method = "euclidean") |> wpgma() |> plot()

```


```{r}

#=====================================================================
# Shape signature results 
#=====================================================================

# The frac option determines the number of bins as a fraction of the average sequence length
# So if average sequence length is 100 and frac is 1/15, then the number of bins is 6.66 --> 7. 
# Each feature histogram contributes one distance metric; 
# the normalize option scales the distances to be between 0 and 1 for every feature.
# This gives each feature equal weight in the distance metric. If not desired, this option
# can be turned off. 

beta_cg <- lapply(beta_seq, seq_to_cgr)  
nadh_cg <- lapply(nadh, seq_to_cgr)  
syn_cg <- lapply(syn_fas, seq_to_cgr) 

cgr_distance2(beta_cg, frac = 1/15, normalize = T) |> hm(trans = scales::modulus_trans(p=2)) 
cgr_distance2(beta_cg, frac = 1/15, normalize = T) |> bionj() |> ape::root("gallus") |> plot()

cgr_distance2(nadh_cg, frac = 1/15, normalize = T) |> hm(trans = scales::modulus_trans(p=2)) 
cgr_distance2(nadh_cg, frac = 1/15, normalize = T) |> bionj() |> root("Saimiri sciureus") |> plot()

cgr_distance2(syn_cg, frac = 1/15, normalize = T) |> hm(trans = scales::modulus_trans(p=2)) 
cgr_distance2(syn_cg, frac = 1/15, normalize = T) |> bionj() |> root("Original") |> plot()

```

```{r, cache = T}

#=====================================================================
# Volume intersection method results 
#=====================================================================

set.seed(0)

#This part of the code takes the longest to run. 
# Reduce samples.per.point and num.points.max to improve runtime or eliminate
# the parameters altogether to use default setting of the hypervolume package, 
# which automatically calculates a number of samples per point based on the 
# dimensions of the data and the number of points.
# For the paper, I used 5000 points, but this method seems to work fine with 
# much fewer points, such as 200, and with the default settings. 


beta_tanimoto <- 
  volume_intersection_tanimoto(
    lapply(beta_cg, function(x) x[-1,]), # Drop the origin point
    hv_args = list(samples.per.point = 5000), 
    vi_args = list(num.points.max = 150000)
  ) 

beta_tanimoto |> (\(x) 1 - x)() |> hm(trans = scales::modulus_trans(p=2)) 
beta_tanimoto |> (\(x) 1 - x)() |> bionj() |> ape::root("gallus") |> plot()

nadh_tanimoto <- 
  volume_intersection_tanimoto(
    lapply(nadh_cg, function(x) x[-1,]), # Drop the origin point
    hv_args = list(samples.per.point = 5000), 
    vi_args = list(num.points.max = 150000)
  ) 

nadh_tanimoto |> (\(x) 1 - x)() |> hm(trans = scales::modulus_trans(p=2)) 
nadh_tanimoto |> (\(x) 1 - x)() |> bionj() |> root("Saimiri sciureus") |> plot()

syn_tanimoto <- 
  volume_intersection_tanimoto(
    lapply(syn_cg, function(x) x[-1,]), # Drop the origin point
    hv_args = list(samples.per.point = 5000), 
    vi_args = list(num.points.max = 150000)) 

syn_tanimoto |> (\(x) 1 - x)() |> hm(trans = scales::modulus_trans(p=2)) 
syn_tanimoto |> (\(x) 1 - x)()  |> bionj() |> plot()

```



```{r}

#=====================================================================
# Demonstrative plots in the paper
#=====================================================================

# Code for 2D chaos game example 
seq_to_2Dcg <- function(dna_seq, plot=F, ...){
  A <- c(0, 0)
  C <- c(0, 1)
  G <- c(1, 1)
  T <- c(1, 0)
  if(length(dna_seq) == 1) dna_seq <- str_split(dna_seq, "")[[1]]
  cg_x <- c(1/2)
  cg_y <- c(1/2)
  for(i in seq_along(dna_seq)){
    cg_x <- c(cg_x, mean(c(cg_x[length(cg_x)], get(dna_seq[i])[1])))
    cg_y <- c(cg_y, mean(c(cg_y[length(cg_y)], get(dna_seq[i])[2])))
  }
  xy <- data.frame(cg_x, cg_y, nucleotide = c("-", dna_seq)) |> 
    rownames_to_column() |> 
    mutate(position = as.numeric(rowname)-1,
           subsequence = "") 
  
  for(i in 2:nrow(xy)){
    xy[["subsequence"]][i] <- paste(xy[["subsequence"]][i-1], xy[["nucleotide"]][i] )
  }
  
  return(xy)
}

# Figure 1.2 An example of 2D chaos game representation for ATGG
rows <- 5
seq_to_2Dcg(beta_seq[[1]]) |> head(rows) |> 
  ggplot(aes(label = subsequence, x = cg_x, y = cg_y)) + 
  geom_text(position = position_nudge(x = -.05)) +
  geom_path(aes(col = as.numeric(rowname)), 
            arrow = arrow(ends = "last", type = "closed", 
                          length = unit(c(rep(.15, rows), 0.25), "inches")), 
            alpha = .75) +
  coord_fixed(xlim = c(0, 1), ylim = c(0, 1)) +
  theme_minimal() + xlab(NULL) + ylab(NULL) +
  guides(col = "none") + 
  annotate(geom = "text", 
           label = c("A", "C", "G", "T"), 
           x = c(0, 0, 1, 1), y = c(0, 1, 1, 0), size = 7) 


# Figure 2.1. 3D CGR of the MT-NADH for the Macaca fascicular
for_plotly <- seq_to_cgr(nadh[[1]])%>% as.data.frame() %>% rownames_to_column() %>% 
    dplyr::rename(position = rowname) %>% mutate(position = as.numeric(position))
fig <- plotly::plot_ly(for_plotly, x = ~i, y = ~j, z = ~k, 
                       size = I(1000), marker = list(line=list(width=0)), color = ~position, opacity = .85)
fig <- fig %>% plotly::add_markers()
fig %>% 
    plotly::config(
        toImageButtonOptions = list(
            format = "png",
            filename = "myplot2",
            width = 1000 * .75,
            height = 750* .75, scale = 2
        )
    )

# Figure 3.2 Sequence alignment of the CDS of theÎ²-globin gene for the chimp, human, and rabbit
alignment_to_vector <- function(alignment){
  output <- alignment$seq
  names(output) <- alignment$nam
  return(output)
}
msa(beta_seq[c("chimp", "human", "rabbit")], type = "dna", method = "ClustalOmega") %>% 
  msaConvert(., type = "seqinr::alignment") |> 
  alignment_to_vector() |> DNAStringSet() |> 
  ggmsa::ggmsa( 1, 52, 
                color = "Shapely_NT", font = "DroidSansMono", char_width = 0.5, 
                seq_name = T, consensus_views = T  ) + 
  theme(axis.text = element_text(size = 20))
msa(beta_seq[c("chimp", "human", "rabbit")], type = "dna", method = "ClustalOmega") %>% 
  msaConvert(., type = "seqinr::alignment") |> 
  alignment_to_vector() |> DNAStringSet() |> 
  ggmsa::ggmsa( 53, 105, 
                color = "Shapely_NT", font = "DroidSansMono", char_width = 0.5,
                seq_name = T, consensus_views = T  ) +
  theme(axis.text = element_text(size = 20))
